spec.bs

```diff
@@ -1902,6 +1935,88 @@ To <dfn>convert a modifier to a string</dfn> given a [=part/modifier=] |modifier
+
+<h3 id=other-specs-json>Integrating with JSON data formats</h3>
+
+JSON data formats which include URL patterns should mirror the behavior of <a href="#other-specs-javascript">JavaScript APIs</a> and accept both:
+* an object which specifies the components required to construct a pattern
+* a string (in the constructor string syntax)
+
+If a specification has an [[INFRA]] value (e.g., after using [=parse a JSON string to an Infra value=]), use the following algorithm, using the appropriate base URL (by default, the URL of the JSON resource).
+
+<div algorithm>
+  To <dfn for=URLPattern>build a {{URLPattern}} from an Infra value</dfn> |rawPattern| given [=/URL=] |baseURL| and [=ECMAScript/realm=] |realm|, perform the following steps.
+
+  1. Let |serializedBaseURL| be the [=URL serializer|serialization=] of |baseURL|.
+  1. If |rawPattern| is a [=string=], then:
+    1. Let |pattern| be a [=new=] {{URLPattern}} with |realm|.
+    1. Run [=initialize=] given |pattern|, |rawPattern|, |serializedBaseURL|, and an empty [=map=].
+
+        <div class="note">It may become necessary in the future to plumb non-empty options here.</div>
+    1. Return |pattern|.
+  1. Otherwise, if |rawPattern| is a [=map=], then:
+    1. Let |init| be «[ "`baseURL`" → |serializedBaseURL| ]», representing a dictionary of type {{URLPatternInit}}.
+    1. [=map/For each=] |key| → |value| of |rawPattern|:
+      1. If |key| is not the <a spec=webidl>identifier</a> of a <a spec=webidl>dictionary member</a> of {{URLPatternInit}} or one of its <a spec=webidl>inherited dictionaries</a>, |value| is not a [=string=], or the member's type is not declared to be {{USVString}}, then return null.
+
+        <div class="note">This will need to be updated if {{URLPatternInit}} gains members of other types.</div>
+        <div class="note">A future version of this specification might also have a less strict mode, if that proves useful to other specifications.</div>
+      1. Set |init|[|key|] to |value|.
+    1. Let |pattern| be a [=new=] {{URLPattern}} with |realm|.
+    1. Run [=initialize=] given |pattern|, |init|, null, and an empty [=map=].
+
+        <div class="note">It may become necessary in the future to plumb non-empty options here.</div>
+    1. Return |pattern|.
+  1. Otherwise, return null.
+
+  <div class="issue">Ideally we wouldn't need a realm here. If we extricate the URL pattern concept from the {{URLPattern}} interface, we won't anymore.</div>
+</div>
+
+Specifications may wish to leave room in their formats to accept options for {{URLPatternOptions}}, override the base URL, or similar, since it is not possible to construct a {{URLPattern}} object directly in this case, unlike in a JavaScript API. For example, [[SPECULATION-RULES]] accepts a "`relative_to`" key which uses the [=document base URL=] instead of the JSON resource's URL.
```

Comment:
```suggestion
Specifications may wish to leave room in their formats to accept options for {{URLPatternOptions}}, override the base URL, or similar, since it is not possible to construct a {{URLPattern}} object directly in this case, unlike in a JavaScript API. For example, [[SPECULATION-RULES]] accepts a "`relative_to`" key which can be used to switch to using the [=document base URL=] instead of the JSON resource's URL.
```

Reply:
Done.